File: academia01/lib/features/pessoas/model/pessoa.dart
//import 'package:flutter/widgets.dart';

enum NivelUsuario {
  admin('ADMIN'),
  user('USER'),
  professor('PROFESSOR'),
  funcionario('FUNCIONARIO'),
  master('MASTER');

  const NivelUsuario(this.value);
  final String value;

  static NivelUsuario fromString(String value) {
    switch (value.toUpperCase()) {
      case 'ADMIN':
        return NivelUsuario.admin;
      case 'USER':
        return NivelUsuario.user;
      case 'PROFESSOR':
        return NivelUsuario.professor;
      case 'FUNCIONARIO':
        return NivelUsuario.funcionario;
      case 'MASTER':
        return NivelUsuario.master;
      default:
        return NivelUsuario.user;
    }
  }
}

abstract class Pessoa {
  final int id;
  final String nome;
  final int idade;
  final String email;
  final String cpf;
  final String senha;
  final NivelUsuario nivel;
  final int idAcademia;
  final int idFuncao;
  final double salario;

  Pessoa({
    required this.id, 
    required this.nome, 
    required this.idade, 
    required this.email, 
    required this.cpf, 
    required this.senha,
    required this.nivel,
    this.idAcademia = 0,
    this.idFuncao = 0,
    this.salario = 0,
  });

  //metodos que devem ser implementados pelas filhas
  Map<String, dynamic> toJson();
  
  //metodos que podem ser sobrescritos
  String get tipoUsuario => nivel.value;

  bool get podeAcessarAcademia => true;

  bool get podeGerenciarAlunos {
    return nivel == NivelUsuario.admin || 
           nivel == NivelUsuario.master || 
           nivel == NivelUsuario.professor;
  }
  
  bool get podeGerenciarFuncionarios {
    return nivel == NivelUsuario.admin || nivel == NivelUsuario.master;
  }
}

File: academia01/lib/services/pessoa.js
class NivelUsuario {
	static ADMIN = 'ADMIN';
	static USER = 'USER';
	static PROFESSOR = 'PROFESSOR';
	static FUNCIONARIO = 'FUNCIONARIO';
	static MASTER = 'MASTER';

	static fromString(value) {
		switch (value.toUpperCase()) {
			case 'ADMIN': return NivelUsuario.ADMIN;
			case 'USER': return NivelUsuario.USER;
			case 'PROFESSOR': return NivelUsuario.PROFESSOR;
			case 'FUNCIONARIO': return NivelUsuario.FUNCIONARIO;
			case 'MASTER': return NivelUsuario.MASTER;
			default: return NivelUsuario.USER;
		}
	}
}

class Pessoa {
	constructor({
		id,
		nome,
		idade,
		email,
		cpf,
		senha,
		nivel,
		idAcademia = 0,
		idFuncao = 0,
		salario = 0,
	}) {
		this.id = id;
		this.nome = nome;
		this.idade = idade;
		this.email = email;
		this.cpf = cpf;
		this.senha = senha;
		this.nivel = nivel;
		this.idAcademia = idAcademia;
		this.idFuncao = idFuncao;
		this.salario = salario;
	}

	toJson() {
		return {
			id: this.id,
			nome: this.nome,
			idade: this.idade,
			email: this.email,
			cpf: this.cpf,
			senha: this.senha,
			nivel: this.nivel,
			idAcademia: this.idAcademia,
			idFuncao: this.idFuncao,
			salario: this.salario,
		};
	}

	get tipoUsuario() {
		return this.nivel;
	}

	get podeAcessarAcademia() {
		return true;
	}

	get podeGerenciarAlunos() {
		return [NivelUsuario.ADMIN, NivelUsuario.MASTER, NivelUsuario.PROFESSOR].includes(this.nivel);
	}

	get podeGerenciarFuncionarios() {
		return [NivelUsuario.ADMIN, NivelUsuario.MASTER].includes(this.nivel);
	}
}

module.exports = { Pessoa, NivelUsuario };
